File: android/src/main/java/me/carda/awesome_notifications/notifications/models/NotificationButtonModel.java
package me.carda.awesome_notifications.notifications.models;

import android.content.Context;

import java.util.Map;
import java.util.HashMap;

import me.carda.awesome_notifications.Definitions;
import me.carda.awesome_notifications.notifications.enumerators.NotificationActionType;
import me.carda.awesome_notifications.utils.StringUtils;
import me.carda.awesome_notifications.notifications.exceptions.AwesomeNotificationException;

public class NotificationButtonModel extends Model {

    public String key;
    public String icon;
    public String label;
    public Boolean enabled;
    public Boolean autoDismissible;
    public Boolean requireInputText;
    public NotificationActionType notificationActionType;

    public NotificationButtonModel(){}

    @Override
    public NotificationButtonModel fromMap(Map<String, Object> arguments) {

        key   = getValueOrDefault(arguments, Definitions.NOTIFICATION_BUTTON_KEY, String.class);
        icon  = getValueOrDefault(arguments, Definitions.NOTIFICATION_BUTTON_ICON, String.class);
        label = getValueOrDefault(arguments, Definitions.NOTIFICATION_BUTTON_LABEL, String.class);

        enabled = getValueOrDefault(arguments, Definitions.NOTIFICATION_ENABLED, Boolean.class);
        autoDismissible = getValueOrDefault(arguments, Definitions.NOTIFICATION_AUTO_DISMISSIBLE, Boolean.class);
        requireInputText = getValueOrDefault(arguments, Definitions.NOTIFICATION_REQUIRE_INPUT_TEXT, Boolean.class);

        notificationActionType = getEnumValueOrDefault(arguments, Definitions.NOTIFICATION_ACTION_TYPE,
                NotificationActionType.class, NotificationActionType.values());

        return this;
    }

    @Override
    public Map<String, Object> toMap() {
        Map<String, Object> returnedObject = new HashMap<>();

        returnedObject.put(Definitions.NOTIFICATION_BUTTON_KEY, key);
        returnedObject.put(Definitions.NOTIFICATION_BUTTON_ICON, icon);
        returnedObject.put(Definitions.NOTIFICATION_BUTTON_LABEL, label);

        returnedObject.put(Definitions.NOTIFICATION_ENABLED, enabled);
        returnedObject.put(Definitions.NOTIFICATION_AUTO_DISMISSIBLE, autoDismissible);
        returnedObject.put(Definitions.NOTIFICATION_REQUIRE_INPUT_TEXT, requireInputText);

        returnedObject.put(Definitions.NOTIFICATION_ACTION_TYPE,
            this.notificationActionType != null ? this.notificationActionType.toString() : NotificationActionType.BringToForeground.toString());

        return returnedObject;
    }

    @Override
    public String toJson() {
        return templateToJson();
    }

    @Override
    public NotificationButtonModel fromJson(String json){
        return (NotificationButtonModel) super.templateFromJson(json);
    }

    @Override
    public void validate(Context context) throws AwesomeNotificationException {
        if(StringUtils.isNullOrEmpty(key))
            throw new AwesomeNotificationException("Button action key cannot be null or empty");

        if(StringUtils.isNullOrEmpty(label))
            throw new AwesomeNotificationException("Button label cannot be null or empty");
    }
}

File: ios/Classes/lib/notifications/models/NotificationButtonModel.swift
//
//  NotificationButtonModel.swift
//  awesome_notifications
//
//  Created by Rafael Setragni on 05/09/20.
//

import Foundation

public class NotificationButtonModel : AbstractModel {
    
    var key:String?
    var icon:String?
    var label:String?

    var enabled:Bool?
    var autoDismissible:Bool?
    var requireInputText:Bool?
    var notificationActionType:NotificationActionType?
    
    public init(){}
    
    public func fromMap(arguments: [String : Any?]?) -> AbstractModel? {
        if(arguments == nil){ return self }
       
        self.key   = MapUtils<String>.getValueOrDefault(reference: Definitions.NOTIFICATION_BUTTON_KEY, arguments: arguments)
        self.icon  = MapUtils<String>.getValueOrDefault(reference: Definitions.NOTIFICATION_BUTTON_ICON, arguments: arguments)
        self.label = MapUtils<String>.getValueOrDefault(reference: Definitions.NOTIFICATION_BUTTON_LABEL, arguments: arguments)

        self.enabled          = MapUtils<Bool>.getValueOrDefault(reference: Definitions.NOTIFICATION_ENABLED, arguments: arguments)
        self.autoDismissible  = MapUtils<Bool>.getValueOrDefault(reference: Definitions.NOTIFICATION_AUTO_DISMISSIBLE, arguments: arguments)
        self.requireInputText = MapUtils<Bool>.getValueOrDefault(reference: Definitions.NOTIFICATION_REQUIRE_INPUT_TEXT, arguments: arguments)
        
        self.notificationActionType = EnumUtils<NotificationActionType>.getEnumOrDefault(
            reference: Definitions.NOTIFICATION_ACTION_TYPE, arguments: arguments)
        
        return self
    }
    
    public func toMap() -> [String : Any?] {
        var mapData:[String: Any?] = [:]
        
        if(key != nil) {mapData[Definitions.NOTIFICATION_BUTTON_KEY] = self.key}
        if(icon != nil) {mapData[Definitions.NOTIFICATION_BUTTON_ICON] = self.icon}
        if(label != nil) {mapData[Definitions.NOTIFICATION_BUTTON_LABEL] = self.label}

        if(enabled != nil) {mapData[Definitions.NOTIFICATION_ENABLED] = self.enabled}
        if(autoDismissible != nil) {mapData[Definitions.NOTIFICATION_AUTO_DISMISSIBLE] = self.autoDismissible}
        if(requireInputText != nil) {mapData[Definitions.NOTIFICATION_REQUIRE_INPUT_TEXT] = self.requireInputText}
        
        if(notificationActionType != nil) {mapData[Definitions.NOTIFICATION_ACTION_TYPE] = self.notificationActionType?.rawValue}
        
        return mapData
    }
    
    public func validate() throws {
        
        if(StringUtils.isNullOrEmpty(key)){
            throw AwesomeNotificationsException.invalidRequiredFields(
                msg: "Button action key cannot be null or empty")
        }

        if(StringUtils.isNullOrEmpty(label)){
            throw AwesomeNotificationsException.invalidRequiredFields(
                msg: "Button label cannot be null or empty")
        }
    }
}
